src/frontend/src/optimizer/plan_node/logical_except.rs

```diff
@@ -0,0 +1,123 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use itertools::Itertools;
+use risingwave_common::error::Result;
+
+use super::{ColPrunable, ExprRewritable, PlanBase, PlanRef, PredicatePushdown, ToBatch, ToStream};
+use crate::optimizer::plan_node::{
+    generic, ColumnPruningContext, LogicalFilter, PlanTreeNode, PredicatePushdownContext,
+    RewriteStreamContext, ToStreamContext,
+};
+use crate::utils::{ColIndexMapping, Condition};
+
+/// `LogicalExcept` returns the rows of its first input except any
+///  matching rows from its other inputs.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalExcept {
+    pub base: PlanBase,
+    core: generic::Except<PlanRef>,
+}
+
+impl LogicalExcept {
+    pub fn new(all: bool, inputs: Vec<PlanRef>) -> Self {
+        let core = generic::Except { all, inputs };
+        let base = PlanBase::new_logical_with_core(&core);
+        LogicalExcept { base, core }
+    }
+
+    pub fn create(all: bool, inputs: Vec<PlanRef>) -> PlanRef {
+        LogicalExcept::new(all, inputs).into()
+    }
+
+    pub(super) fn fmt_with_name(&self, f: &mut fmt::Formatter<'_>, name: &str) -> fmt::Result {
+        self.core.fmt_with_name(f, name)
+    }
+
+    pub fn fmt_fields_with_builder(&self, builder: &mut fmt::DebugStruct<'_, '_>) {
+        self.core.fmt_fields_with_builder(builder)
+    }
+
+    pub fn all(&self) -> bool {
+        self.core.all
+    }
+}
+
+impl PlanTreeNode for LogicalExcept {
+    fn inputs(&self) -> smallvec::SmallVec<[crate::optimizer::PlanRef; 2]> {
+        let mut vec = smallvec::SmallVec::new();
+        vec.extend(self.core.inputs.clone().into_iter());
+        vec
+    }
+
+    fn clone_with_inputs(&self, inputs: &[crate::optimizer::PlanRef]) -> PlanRef {
+        Self::new(self.all(), inputs.to_vec()).into()
+    }
+}
+
+impl fmt::Display for LogicalExcept {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.fmt_with_name(f, "LogicalExcept")
+    }
+}
+
+impl ColPrunable for LogicalExcept {
+    fn prune_col(&self, required_cols: &[usize], ctx: &mut ColumnPruningContext) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.prune_col(required_cols, ctx))
+            .collect_vec();
+        self.clone_with_inputs(&new_inputs)
+    }
+}
+
+impl ExprRewritable for LogicalExcept {}
+
+impl PredicatePushdown for LogicalExcept {
+    fn predicate_pushdown(
+        &self,
+        predicate: Condition,
+        ctx: &mut PredicatePushdownContext,
+    ) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.predicate_pushdown(Condition::true_cond(), ctx))
```

Comment:
I guess we can push down predicates to all inputs?

Reply:
Yes, I was too conservative before, because `predicate push down` will be applied after `SetOperatorToJoin`. I think predicates can be pushed down through all set operations.